File: 695-max-area-of-island/695-max-area-of-island.rb
# @param {Integer[][]} grid
# @return {Integer}

def in_bounds(row, col, grid)
    return row >= 0 && row < grid.length && col >= 0 && col < grid[0].length
end

def valid(row, col, grid, visited)
    return in_bounds(row, col, grid) && !visited.include?([row, col]) && grid[row][col] == 1
end

def dfs(row, col, grid, visited)
    return 0 if !valid(row, col, grid, visited)
    
    visited.add([row, col])
    dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]]
    return dirs.reduce(1) { |sum, dir| sum + dfs(row+dir[0], col+dir[1], grid, visited) }
end

def max_area_of_island(grid)
    visited = Set.new
    max_size = 0
    
    (0...grid.length).each do |row|
        (0...grid[row].length).each do |col|
            if valid(row, col, grid, visited)
                max_size = [max_size, dfs(row, col, grid, visited)].max
            end
        end
    end
    
    return max_size
end

File: 695-max-area-of-island/695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        visited = set()
        maxSize = 0
        
        def inBounds(row, col):
            return row in range(ROWS) and col in range(COLS)
        
        def valid(row, col):
            return inBounds(row, col) and (row, col) not in visited and grid[row][col] == 1
        
        def dfs(row, col):
            if not valid(row, col):
                return 0
            else:
                visited.add((row, col))
                return 1 + dfs(row+1, col) + dfs(row-1, col) + dfs(row, col+1) + dfs(row, col-1)
            
        for row in range(ROWS):
            for col in range(COLS):
                if valid(row, col):
                    maxSize = max(maxSize, dfs(row, col))
                    
        return maxSize